File: packages/template-ts/src/functions/date.ts
// date and now (FUN-37 to FUN-42).
import { isString, textOf, type Value } from '../value/value.js';
import { argString, typeError, type BuiltIn, type FunctionContext } from './helpers.js';

const DAY_SHORT = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
const DAY_LONG = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
const MONTH_SHORT = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
const MONTH_LONG = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];

// Offset in seconds of `Z` or `±HH:MM`, or null when the text is not an offset.
export function parseOffset(text: string): number | null {
  if (text === 'Z') return 0;
  const match = /^([+-])(\d{2}):(\d{2})$/.exec(text);
  if (!match) return null;
  const sign = match[1] === '-' ? -1 : 1;
  const hours = Number(match[2]);
  const minutes = Number(match[3]);
  if (hours > 23 || minutes > 59) return null;
  return sign * (hours * 3600 + minutes * 60);
}

// Days since 1970-01-01 of a proleptic Gregorian date.
function daysFromCivil(year: number, month: number, day: number): number {
  const y = month <= 2 ? year - 1 : year;
  const era = Math.floor(y / 400);
  const yoe = y - era * 400;
  const mp = (month + 9) % 12;
  const doy = Math.floor((153 * mp + 2) / 5) + day - 1;
  const doe = yoe * 365 + Math.floor(yoe / 4) - Math.floor(yoe / 100) + doy;
  return era * 146097 + doe - 719468;
}

function civilFromDays(days: number): { year: number; month: number; day: number } {
  const z = days + 719468;
  const era = Math.floor(z / 146097);
  const doe = z - era * 146097;
  const yoe = Math.floor((doe - Math.floor(doe / 1460) + Math.floor(doe / 36524) - Math.floor(doe / 146096)) / 365);
  const y = yoe + era * 400;
  const doy = doe - (365 * yoe + Math.floor(yoe / 4) - Math.floor(yoe / 100));
  const mp = Math.floor((5 * doy + 2) / 153);
  const day = doy - Math.floor((153 * mp + 2) / 5) + 1;
  const month = mp < 10 ? mp + 3 : mp - 9;
  return { year: month <= 2 ? y + 1 : y, month, day };
}

const DATE_TEXT = /^(\d{4})-(\d{2})-(\d{2})(?:[ T](\d{2}):(\d{2}):(\d{2}))?(Z|[+-]\d{2}:\d{2})?$/;

// Unix seconds of a date value (FUN-37).
export function toUnixSeconds(value: Value, envOffset: number): number {
  if (typeof value === 'number') return Math.trunc(value);
  if (isString(value)) {
    const match = DATE_TEXT.exec(textOf(value));
    if (!match) throw typeError(`${JSON.stringify(textOf(value))} is not a date`);
    const year = Number(match[1]);
    const month = Number(match[2]);
    const day = Number(match[3]);
    const hour = match[4] === undefined ? 0 : Number(match[4]);
    const minute = match[5] === undefined ? 0 : Number(match[5]);
    const second = match[6] === undefined ? 0 : Number(match[6]);
    if (month < 1 || month > 12 || day < 1 || day > 31 || hour > 23 || minute > 59 || second > 59) {
      throw typeError(`${JSON.stringify(textOf(value))} is not a date`);
    }
    const offset = match[7] === undefined ? envOffset : (parseOffset(match[7]) as number);
    return daysFromCivil(year, month, day) * 86400 + hour * 3600 + minute * 60 + second - offset;
  }
  throw typeError('date requires a number or a string');
}

function pad(value: number, width: number): string {
  return String(value).padStart(width, '0');
}

export function formatDate(seconds: number, format: string, offset: number): string {
  const local = seconds + offset;
  const days = Math.floor(local / 86400);
  const secondOfDay = local - days * 86400;
  const { year, month, day } = civilFromDays(days);
  const hour = Math.floor(secondOfDay / 3600);
  const minute = Math.floor((secondOfDay % 3600) / 60);
  const second = secondOfDay % 60;
  const weekday = ((days % 7) + 11) % 7; // 1970-01-01 is a Thursday (4).
  const sign = offset < 0 ? '-' : '+';
  const absOffset = Math.abs(offset);
  let result = '';
  for (let i = 0; i < format.length; i++) {
    const char = format[i] as string;
    switch (char) {
      case '\\': result += format[i + 1] ?? ''; i++; break;
      case 'Y': result += pad(year, 4); break;
      case 'y': result += pad(year % 100, 2); break;
      case 'm': result += pad(month, 2); break;
      case 'n': result += String(month); break;
      case 'd': result += pad(day, 2); break;
      case 'j': result += String(day); break;
      case 'H': result += pad(hour, 2); break;
      case 'G': result += String(hour); break;
      case 'i': result += pad(minute, 2); break;
      case 's': result += pad(second, 2); break;
      case 'D': result += DAY_SHORT[weekday] as string; break;
      case 'l': result += DAY_LONG[weekday] as string; break;
      case 'N': result += String(weekday === 0 ? 7 : weekday); break;
      case 'w': result += String(weekday); break;
      case 'M': result += MONTH_SHORT[month - 1] as string; break;
      case 'F': result += MONTH_LONG[month - 1] as string; break;
      case 'U': result += String(seconds); break;
      case 'P': result += `${sign}${pad(Math.floor(absOffset / 3600), 2)}:${pad(Math.floor((absOffset % 3600) / 60), 2)}`; break;
      default: result += char;
    }
  }
  return result;
}

function envOffset(context: FunctionContext): number {
  const offset = parseOffset(context.env.timezone);
  if (offset === null) throw typeError(`${JSON.stringify(context.env.timezone)} is not a time zone offset`);
  return offset;
}

export const dateFunctions: Record<string, BuiltIn> = {
  date: {
    min: 2,
    max: 2,
    call: ([v, fmt], context) => {
      if (v === null) return '';
      const offset = envOffset(context);
      return formatDate(toUnixSeconds(v as Value, offset), argString(fmt as Value, 'date'), offset);
    },
  },
  now: { min: 0, max: 0, call: (_args, context) => context.env.now },
};

File: packages/template-rust/src/functions/date.rs
//! date and now (FUN-37 to FUN-42).

use crate::functions::helpers::{BuiltIn, FunctionContext, FunctionError, arg_string, type_error};
use crate::value::Value;

const DAY_SHORT: [&str; 7] = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
const DAY_LONG: [&str; 7] = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"];
const MONTH_SHORT: [&str; 12] = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
const MONTH_LONG: [&str; 12] = [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
];

/// Offset in seconds of `Z` or `±HH:MM`, or None.
pub fn parse_offset(text: &str) -> Option<i64> {
    if text == "Z" {
        return Some(0);
    }
    let bytes = text.as_bytes();
    if bytes.len() != 6 || !(bytes[0] == b'+' || bytes[0] == b'-') || bytes[3] != b':' {
        return None;
    }
    let hours: i64 = text[1..3].parse().ok()?;
    let minutes: i64 = text[4..6].parse().ok()?;
    if !text[1..3].bytes().all(|b| b.is_ascii_digit()) || !text[4..6].bytes().all(|b| b.is_ascii_digit()) || hours > 23 || minutes > 59 {
        return None;
    }
    let sign = if bytes[0] == b'-' { -1 } else { 1 };
    Some(sign * (hours * 3600 + minutes * 60))
}

fn days_from_civil(year: i64, month: i64, day: i64) -> i64 {
    let y = if month <= 2 { year - 1 } else { year };
    let era = y.div_euclid(400);
    let yoe = y - era * 400;
    let mp = (month + 9) % 12;
    let doy = (153 * mp + 2) / 5 + day - 1;
    let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
    era * 146097 + doe - 719468
}

fn civil_from_days(days: i64) -> (i64, i64, i64) {
    let z = days + 719468;
    let era = z.div_euclid(146097);
    let doe = z - era * 146097;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let day = doy - (153 * mp + 2) / 5 + 1;
    let month = if mp < 10 { mp + 3 } else { mp - 9 };
    (if month <= 2 { y + 1 } else { y }, month, day)
}

/// Fields of a parsed date text: year, month, day, hour, minute, second and the offset text.
struct DateText<'a> {
    year: i64,
    month: i64,
    day: i64,
    hour: i64,
    minute: i64,
    second: i64,
    offset: Option<&'a str>,
}

fn parse_date_text(text: &str) -> Option<DateText<'_>> {
    let bytes = text.as_bytes();
    if bytes.len() < 10 || bytes[4] != b'-' || bytes[7] != b'-' {
        return None;
    }
    let digits = |range: std::ops::Range<usize>| -> Option<i64> {
        let part = text.get(range)?;
        if part.bytes().all(|b| b.is_ascii_digit()) {
            part.parse().ok()
        } else {
            None
        }
    };
    let year = digits(0..4)?;
    let month = digits(5..7)?;
    let day = digits(8..10)?;
    let mut rest = &text[10..];
    let (mut hour, mut minute, mut second) = (0, 0, 0);
    if rest.starts_with(' ') || rest.starts_with('T') {
        if rest.len() < 9 || rest.as_bytes()[3] != b':' || rest.as_bytes()[6] != b':' {
            return None;
        }
        let time = &rest[1..9];
        let time_digits = |range: std::ops::Range<usize>| -> Option<i64> {
            let part = time.get(range)?;
            if part.bytes().all(|b| b.is_ascii_digit()) {
                part.parse().ok()
            } else {
                None
            }
        };
        hour = time_digits(0..2)?;
        minute = time_digits(3..5)?;
        second = time_digits(6..8)?;
        rest = &rest[9..];
    }
    let offset = if rest.is_empty() { None } else { Some(rest) };
    if let Some(offset) = offset {
        parse_offset(offset)?;
    }
    Some(DateText {
        year,
        month,
        day,
        hour,
        minute,
        second,
        offset,
    })
}

/// Unix seconds of a date value (FUN-37).
pub fn to_unix_seconds(value: &Value, env_offset: i64) -> Result<i64, FunctionError> {
    match value {
        Value::Number(number) => Ok(number.trunc() as i64),
        Value::Str(text) | Value::Safe(text) => {
            let Some(DateText {
                year,
                month,
                day,
                hour,
                minute,
                second,
                offset,
            }) = parse_date_text(text)
            else {
                return Err(type_error(format!("{text:?} is not a date")));
            };
            if !(1..=12).contains(&month) || !(1..=31).contains(&day) || hour > 23 || minute > 59 || second > 59 {
                return Err(type_error(format!("{text:?} is not a date")));
            }
            let offset = offset.and_then(parse_offset).unwrap_or(env_offset);
            Ok(days_from_civil(year, month, day) * 86400 + hour * 3600 + minute * 60 + second - offset)
        }
        _ => Err(type_error("date requires a number or a string")),
    }
}

/// Formats unix seconds with the tokens of FUN-40 in a fixed offset.
pub fn format_date(seconds: i64, format: &str, offset: i64) -> String {
    let local = seconds + offset;
    let days = local.div_euclid(86400);
    let second_of_day = local - days * 86400;
    let (year, month, day) = civil_from_days(days);
    let hour = second_of_day / 3600;
    let minute = (second_of_day % 3600) / 60;
    let second = second_of_day % 60;
    let weekday = ((days % 7) + 11).rem_euclid(7) as usize;
    let sign = if offset < 0 { '-' } else { '+' };
    let abs_offset = offset.abs();
    let mut result = String::new();
    let chars: Vec<char> = format.chars().collect();
    let mut i = 0;
    while i < chars.len() {
        match chars[i] {
            '\\' => {
                if let Some(next) = chars.get(i + 1) {
                    result.push(*next);
                }
                i += 1;
            }
            'Y' => result.push_str(&format!("{year:04}")),
            'y' => result.push_str(&format!("{:02}", year.rem_euclid(100))),
            'm' => result.push_str(&format!("{month:02}")),
            'n' => result.push_str(&month.to_string()),
            'd' => result.push_str(&format!("{day:02}")),
            'j' => result.push_str(&day.to_string()),
            'H' => result.push_str(&format!("{hour:02}")),
            'G' => result.push_str(&hour.to_string()),
            'i' => result.push_str(&format!("{minute:02}")),
            's' => result.push_str(&format!("{second:02}")),
            'D' => result.push_str(DAY_SHORT[weekday]),
            'l' => result.push_str(DAY_LONG[weekday]),
            'N' => result.push_str(&(if weekday == 0 { 7 } else { weekday }).to_string()),
            'w' => result.push_str(&weekday.to_string()),
            'M' => result.push_str(MONTH_SHORT[(month - 1) as usize]),
            'F' => result.push_str(MONTH_LONG[(month - 1) as usize]),
            'U' => result.push_str(&seconds.to_string()),
            'P' => result.push_str(&format!("{sign}{:02}:{:02}", abs_offset / 3600, (abs_offset % 3600) / 60)),
            other => result.push(other),
        }
        i += 1;
    }
    result
}

fn env_offset(context: &FunctionContext<'_>) -> Result<i64, FunctionError> {
    parse_offset(&context.env.timezone).ok_or_else(|| type_error(format!("{:?} is not a time zone offset", context.env.timezone)))
}

fn date(args: &[Value], context: &FunctionContext<'_>) -> Result<Value, FunctionError> {
    if args[0] == Value::Null {
        return Ok(Value::text(String::new()));
    }
    let offset = env_offset(context)?;
    let seconds = to_unix_seconds(&args[0], offset)?;
    Ok(Value::text(format_date(seconds, &arg_string(&args[1], "date")?, offset)))
}

fn now(_args: &[Value], context: &FunctionContext<'_>) -> Result<Value, FunctionError> {
    Ok(Value::Number(context.env.now))
}

/// The time function group.
pub fn functions() -> Vec<(&'static str, BuiltIn)> {
    vec![
        (
            "date",
            BuiltIn {
                min: 2,
                max: 2,
                call: date,
            },
        ),
        ("now", BuiltIn { min: 0, max: 0, call: now }),
    ]
}
